{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "40952c2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load required R packages\n",
    "# 载入需要的R包\n",
    "library(tidyverse)\n",
    "library(DESeq2)\n",
    "library(BioSeqUtils)\n",
    "library(caret)\n",
    "library(randomForest)\n",
    "library(ROCR) \n",
    "library(pROC)\n",
    "library(ggplot2)\n",
    "library(PRROC)\n",
    "library(tidymodels)\n",
    "library(Seurat)\n",
    "library(patchwork)\n",
    "library(ggsci)\n",
    "library(ada)\n",
    "library(plotrix)\n",
    "library(kernlab)\n",
    "library(precrec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a58a34ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "#setting working directory\n",
    "# 设置工作路径\n",
    "setwd(\"D:\\\\project\\\\白骨壤Cd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "511c2cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "theme_cleveland <- function (rotate = TRUE) \n",
    "{\n",
    "    if (rotate) {\n",
    "        theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(), \n",
    "            panel.grid.major.y = element_line(colour = \"grey70\", \n",
    "                linetype = \"dashed\"))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1802e7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model evaluation metrics\n",
    "# 模型评估指标\n",
    "metrics2 <- function(Pred,realY){\n",
    "  AUC_ROC <- roc(realY,Pred)\n",
    "  AUC_ROC <- AUC_ROC$auc\n",
    "  pred.class <- as.integer(Pred > 0.5)\n",
    "  TBL <- table(pred.class, realY)\n",
    "  TP <- TBL[2,2]\n",
    "  FP <- TBL[2,1]                  \n",
    "  TN <- TBL[1,1]                                                                     \n",
    "  FN <- TBL[1,2]\n",
    "  Specificity <- TN/(TN+FP)\n",
    "  Sensitivity <- TP/(TP+FN)\n",
    "  Precision <- TP/(TP+FP)\n",
    "  Recall <- TP/(TP+FN)\n",
    "  F1 <- 2 * (Precision*Recall/(Precision+Recall))\n",
    "  MCC <- (TP*TN-FP*FN)/sqrt((TP+FP)*(TP+FN))/sqrt((TN+FP)*(TN+FN))\n",
    "  Accuracy <- (TP+TN)/(TP+TN+FP+FN)\n",
    "  tmp <- data.frame(V1=realY, V2=Pred)\n",
    "  tmp$V1 <- factor(tmp$V1)\n",
    "  tmp <- evalmod(scores = tmp$V2, labels = tmp$V1)\n",
    "  aucs <- auc(tmp)\n",
    "  AUC_PRC <- subset(aucs, curvetypes == \"PRC\")\n",
    "  AUC_PRC <- AUC_PRC$aucs\n",
    "  re <- data.frame(Sensitivity,Specificity,Precision,Recall,Accuracy,AUC_ROC,AUC_PRC,MCC,F1)\n",
    "  return(re)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "19c4412a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom function for differentially expressed genes analysis using DESeq2 R package\n",
    "# 构建函数，使用DESeq2进行差异表达分析\n",
    "DEG <- function(count,sampleNames,comparison,rep=3){\n",
    "    data <- count[,sampleNames]\n",
    "    countData <- as.matrix(data)\n",
    "    condition <- c(rep(\"control\",rep),rep(\"treatment\",rep))\n",
    "    database <- data.frame(name=sampleNames, condition=condition)\n",
    "    database$condition <- factor(database$condition)\n",
    "    rownames(database) <- sampleNames\n",
    "    dds <- DESeqDataSetFromMatrix(countData, colData=database, design= ~ condition)\n",
    "    dds <- dds[ rowSums(counts(dds)) > 10, ]\n",
    "    dds <- DESeq(dds)\n",
    "    res <- results(dds,contrast=c(\"condition\", \"treatment\", \"control\"))\n",
    "    res <- res %>%\n",
    "        as.data.frame() %>%\n",
    "        mutate(comparison=comparison,gene=rownames(res))\n",
    "    return(res)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5668b5b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom function for reading gene expression in raw counts format\n",
    "# 读取基因表达数据(格式raw counts)\n",
    "read_count <- function(path=\"03_raw_counts/\",file=\"Avicennia_marina.counts\"){\n",
    "    Count <- read.table(paste0(path,file),head=T,check.names = F,row.names=1)\n",
    "    Count <- Count %>%\n",
    "        dplyr::select(-c(Chr,Start,End,Strand,Length))\n",
    "        Cols <- lapply(colnames(Count),function(x){\n",
    "            unlist(str_split(x,'[-.]'))[3]\n",
    "        }) \n",
    "    colnames(Count) <- unlist(Cols)\n",
    "    return(Count)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "453787fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom function for reading gene expression in FPKM format\n",
    "# 读取基因表达数据(格式FPKM)\n",
    "read_fpkm <- function(path=\"04_TPM_FPKM/\",file=\"Athaliana.fpkm\"){\n",
    "    fpkm <- read.table(paste0(path,file),head=T,check.names = F,row.names=1)\n",
    "        Cols <- lapply(colnames(fpkm),function(x){\n",
    "            unlist(str_split(x,'[-.]'))[3]\n",
    "        }) \n",
    "    colnames(fpkm) <- unlist(Cols)\n",
    "    return(fpkm)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8c17344f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Count <- read_count(path=\"03_raw_counts/\",file=\"Avicennia_marina.counts\")\n",
    "fpkm <- read_fpkm(path=\"04_TPM_FPKM//\",file=\"Avicennia_marina.fpkm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8767d798",
   "metadata": {},
   "outputs": [],
   "source": [
    "fpkm <- fpkm[,c(\"Con_2\",\"Con_3\",\"Con_4\",\"Cd_1\",'Cd_2','Cd_4')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5e98c772",
   "metadata": {},
   "outputs": [],
   "source": [
    "fpkm <- data.frame(gene=rownames(fpkm),avg_fpkm=apply(fpkm,1,mean))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "948a3355",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Am_deg <- DEG(count=Count,\n",
    "    sampleNames= c(\"Con_2\",\"Con_3\",\"Con_4\",\"Cd_1\",'Cd_2','Cd_4'),\n",
    "    comparison=c(\"Cd\"),\n",
    "          rep=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "35ff1fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Am_deg <- Am_deg[!is.na(Am_deg$padj),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f2f8014e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading genomics sequence features\n",
    "# 读取基因特征矩阵\n",
    "Am <- read.csv(\"期刊/JHM/revise/Avicennia_marina_feature.csv\")\n",
    "colnames(Am)[1] <- 'gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c0750034",
   "metadata": {},
   "outputs": [],
   "source": [
    "Am2 <- Am %>%\n",
    "    dplyr::filter(gene %in% Am_deg$gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ca1c8385",
   "metadata": {},
   "outputs": [],
   "source": [
    "Am_DEG <- Am_deg\n",
    "Am_DEG <- Am_DEG %>%\n",
    "    mutate(Organism = \"Avicennia marina\") %>%\n",
    "    dplyr::filter(gene %in% Am2$gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1b345f75",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Data is of class data.frame. Coercing to dgCMatrix.\"\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  T.Intron, G.Intron, A.Intron, C.Intron, TT.Intron, AT.Intron, TG.Intron, CT.Intron, GT.Intron, GA.Intron \n",
      "\t   AG.Intron, TA.Intron, TC.Intron, CA.Intron, AA.Intron, AC.Intron, GC.Intron, GG.Intron, CC.Intron, CG.Intron \n",
      "\t   TA.CDS, AT.CDS, AA.CDS, GG.5UTR, TT.CDS, CC.3UTR, CG.5UTR, TG.CDS, GC.3UTR, CG.Down \n",
      "Negative:  G.CDS, C.CDS, T.3UTR, GC.CDS, CG.CDS, GG.CDS, A.Down, T.Up, CC.CDS, T.Down \n",
      "\t   A.3UTR, TC.CDS, A.Up, GT.Down, AC.Up, C.5UTR, C.Up, CA.Up, TG.Down, GT.Up \n",
      "\t   CA.5UTR, TT.3UTR, TC.Up, G.Up, G.Down, AT.3UTR, TG.Up, GA.Down, AC.Down, A.5UTR \n",
      "PC_ 2 \n",
      "Positive:  GC.Down, C.Down, CC.Down, G.Down, GG.Down, GC.Up, GG.Up, G.Up, CT.Down, AG.Down \n",
      "\t   CA.Down, TC.Down, CC.Up, C.Up, CG.Down, GA.Down, C.3UTR, CT.Up, CG.Up, TG.Up \n",
      "\t   G.5UTR, AG.Up, AC.Down, GA.Up, GC.3UTR, TG.5UTR, TC.Up, GG.5UTR, TG.Down, CC.3UTR \n",
      "Negative:  TA.Down, AT.Down, TA.Up, TT.Down, AT.Up, T.Down, AA.Down, A.Down, A.Up, AA.Up \n",
      "\t   TA.3UTR, TT.Up, T.Up, AT.3UTR, T.3UTR, CG.CDS, TT.3UTR, TA.5UTR, A.5UTR, AT.5UTR \n",
      "\t   C.CDS, AA.5UTR, CC.CDS, GC.CDS, AA.3UTR, A.3UTR, GG.CDS, G.CDS, AC.5UTR, TC.CDS \n",
      "PC_ 3 \n",
      "Positive:  GC.Up, G.Up, GG.Up, C.Up, CC.Up, TA.Down, AT.Down, CT.Up, TG.Up, CG.Up \n",
      "\t   AG.Up, TC.Up, GA.Up, AA.Down, CA.Up, TT.Down, GT.Up, A.Down, T.Down, AC.Up \n",
      "\t   G.5UTR, GG.5UTR, TG.5UTR, AG.5UTR, GC.5UTR, GA.5UTR, TA.3UTR, AT.3UTR, GT.5UTR, CT.5UTR \n",
      "Negative:  TA.Up, AT.Up, GC.Down, AA.Up, G.Down, C.Down, A.Up, CC.Down, GG.Down, TT.Up \n",
      "\t   T.Up, CT.Down, AG.Down, TC.Down, GA.Down, CG.Down, CA.Down, TG.Down, AC.Down, TA.5UTR \n",
      "\t   AT.5UTR, GT.Down, C.3UTR, AA.5UTR, A.5UTR, GC.3UTR, G.3UTR, CT.3UTR, CC.3UTR, T.5UTR \n",
      "PC_ 4 \n",
      "Positive:  AT.CDS, TA.CDS, T.CDS, TG.CDS, TG.5UTR, TT.CDS, A.CDS, GT.5UTR, TT.5UTR, T.5UTR \n",
      "\t   AA.CDS, G.5UTR, GG.5UTR, CT.3UTR, TG.3UTR, T.Down, TT.Down, TT.Up, TA.Down, AT.Down \n",
      "\t   GC.3UTR, G.3UTR, T.Up, GA.5UTR, C.3UTR, GA.CDS, GT.CDS, GT.3UTR, GG.3UTR, AG.CDS \n",
      "Negative:  CG.CDS, C.CDS, CC.CDS, GC.CDS, AC.5UTR, CA.5UTR, CC.5UTR, C.5UTR, AC.CDS, CG.Intron \n",
      "\t   AA.3UTR, A.5UTR, CC.Down, C.Down, A.3UTR, AA.5UTR, CA.Down, TC.CDS, GC.Down, AC.Down \n",
      "\t   TA.3UTR, CG.Down, CA.Up, AT.3UTR, AC.Up, GG.CDS, AG.Down, CC.Up, C.Up, GG.Down \n",
      "PC_ 5 \n",
      "Positive:  TA.5UTR, AT.5UTR, A.5UTR, AA.5UTR, T.5UTR, TT.5UTR, TG.Up, G.Up, GG.Up, T.CDS \n",
      "\t   GT.Up, GC.Up, AG.Up, TA.CDS, GA.Up, TT.CDS, CT.Up, AT.3UTR, AT.CDS, TA.3UTR \n",
      "\t   CT.CDS, A.3UTR, C.Up, AA.3UTR, TC.Up, CC.Up, T.3UTR, TC.CDS, TT.3UTR, CG.Up \n",
      "Negative:  GC.5UTR, C.5UTR, CG.5UTR, CC.5UTR, G.5UTR, TC.5UTR, GG.5UTR, CT.5UTR, AG.5UTR, GA.5UTR \n",
      "\t   GT.5UTR, CA.5UTR, AC.5UTR, TG.5UTR, TA.Up, AT.Up, AA.Up, GG.CDS, A.Up, G.3UTR \n",
      "\t   G.CDS, GG.3UTR, CG.3UTR, GC.3UTR, AG.CDS, C.3UTR, GA.CDS, CC.3UTR, GC.CDS, CG.CDS \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 25527\n",
      "Number of edges: 650309\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.7283\n",
      "Number of communities: 11\n",
      "Elapsed time: 5 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session\"\n",
      "14:58:50 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "14:58:50 Read 25527 rows and found 10 numeric columns\n",
      "\n",
      "14:58:50 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "14:58:50 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "14:58:53 Writing NN index file to temp file C:\\Users\\jasonxu\\AppData\\Local\\Temp\\Rtmpu2o3YY\\file884843e7367b\n",
      "\n",
      "14:58:53 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "14:59:02 Annoy recall = 100%\n",
      "\n",
      "14:59:02 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "14:59:04 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "14:59:04 Using 'irlba' for PCA\n",
      "\n",
      "14:59:04 PCA: 2 components explained 49.58% variance\n",
      "\n",
      "14:59:04 Scaling init to sdev = 1\n",
      "\n",
      "14:59:04 Commencing optimization for 200 epochs, with 954498 positive edges\n",
      "\n",
      "14:59:04 Using rng type: pcg\n",
      "\n",
      "14:59:26 Optimization finished\n",
      "\n"
     ]
    }
   ],
   "source": [
    "rownames(Am2) <- Am2$gene\n",
    "Am2 <- Am2[,-1]\n",
    "Am2 <- data.frame(t(Am2))\n",
    "objs <- CreateSeuratObject(counts = Am2, project = \"all\", meta=Am_deg,min.cells = 1, min.features = 1)\n",
    "objs <- NormalizeData(objs, normalization.method = \"LogNormalize\", scale.factor = 30)\n",
    "objs <- FindVariableFeatures(objs, selection.method = \"vst\", nfeatures = 3000)\n",
    "all.genes <- rownames(objs)\n",
    "objs <- ScaleData(objs, features = all.genes)\n",
    "objs <- RunPCA(objs,npcs =10, features = VariableFeatures(object = objs))\n",
    "objs <- FindNeighbors(objs, dims = 1:10)\n",
    "objs <- FindClusters(objs, resolution = 1)\n",
    "objs <- RunUMAP(objs, dims = 1:10)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "234dfab3",
   "metadata": {},
   "source": [
    "## not run, this step including negative samples sampling and other machine learning models training\n",
    "# 这一步包含使用先聚类后对负样本进行采集，使用其他机器学习模型进行模型训练\n",
    "for(ML in c('rpart','ada','lda','knn','glm','svmRadial','naive_bayes','xgbTree','treebag')){\n",
    "    print(ML)\n",
    "    TBL <- NULL\n",
    "    for(lfc in c(0.5,1,1.5,2)){\n",
    "        print(lfc)\n",
    "        RG <- Am_deg %>%\n",
    "                dplyr::filter(abs(log2FoldChange) > lfc & padj < 0.05) %>%\n",
    "                dplyr::select(gene) %>%\n",
    "                mutate(class=1)\n",
    "        NRG <- Am_deg %>%\n",
    "                dplyr::filter(abs(log2FoldChange) < 0.5) %>%\n",
    "                dplyr::select(gene) %>%\n",
    "                mutate(class=0)\n",
    "        genes <- rbind(RG,NRG)\n",
    "\n",
    "        meta <- objs@meta.data %>%\n",
    "                    inner_join(genes)\n",
    "        tbl <- table(meta[,c(\"seurat_clusters\",\"class\")])\n",
    "        tbl <- tbl %>%\n",
    "            data.frame() %>%\n",
    "            mutate(class=ifelse(class=='0','NRG','RG')) %>%\n",
    "            spread(.,class,Freq) %>%\n",
    "            arrange(RG) %>%\n",
    "            mutate(log2FC=lfc)\n",
    "        TBL <- rbind(TBL,tbl)\n",
    "        Number <- 0\n",
    "        inx <- c()\n",
    "        for(i in 1:nrow(tbl)){\n",
    "            print(tbl$seurat_clusters[i])\n",
    "            re <- tbl$NRG[i]\n",
    "            Number <- Number + re\n",
    "            if(Number < nrow(RG)){\n",
    "                inx <- c(inx,i)\n",
    "            }else{\n",
    "                inx <- c(inx,i)\n",
    "                break\n",
    "            }\n",
    "        }\n",
    "        need <- meta %>%\n",
    "            dplyr::filter(seurat_clusters %in% tbl$seurat_clusters[inx]) %>%\n",
    "            dplyr::filter(class==0) %>%\n",
    "            dplyr::select(gene,class)\n",
    "        set.seed(123)\n",
    "        inx <- sample(1:nrow(need),nrow(RG))\n",
    "        NRG <- need[inx,]\n",
    "        genes <- rbind(RG,NRG)\n",
    "        new_data <-  Am_deg %>%\n",
    "            inner_join(genes) %>%\n",
    "            dplyr::select(gene,class) %>%\n",
    "            inner_join(fpkm) %>%\n",
    "            dplyr::select(gene,avg_fpkm,class)\n",
    "        data <- new_data %>%\n",
    "            inner_join(Am)\n",
    "        rownames(data) <- data$gene\n",
    "        data <- data %>%\n",
    "                    mutate(class=ifelse(class==0,'NRG','RG')) %>%\n",
    "                    dplyr::select(-c(avg_fpkm,gene))\n",
    "        ctrl <- trainControl(method = \"cv\",number=10)\n",
    "        ML_feature <- as.integer(sqrt(ncol(data)-1))\n",
    "        grid_search = expand.grid(.mtry=ML_feature)\n",
    "        set.seed(123)\n",
    "        inx <- sample(1:nrow(data),0.75*nrow(data))\n",
    "        train <- data[inx,]\n",
    "        test <- data[-inx,]\n",
    "        set.seed(123)\n",
    "        model <- train(class ~ ., data = train,\n",
    "                    method = ML,\n",
    "                    trControl = ctrl)\n",
    "        save(train,test,model,file=paste0(\"model/\",ML,\"/\",\"clustering\",\"_log2FC_\",lfc,\".RData\"))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ca3a7f1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "abe8e5cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rf\"\n",
      "[1] \"model/rf/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.851\"\n",
      "[1] \"model/rf/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.971\"\n",
      "[1] \"model/rf/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.989\"\n",
      "[1] \"model/rf/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.968\"\n",
      "[1] \"rpart\"\n",
      "[1] \"model/rpart/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.718\"\n",
      "[1] \"model/rpart/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.867\"\n",
      "[1] \"model/rpart/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.873\"\n",
      "[1] \"model/rpart/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.861\"\n",
      "[1] \"ada\"\n",
      "[1] \"model/ada/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.825\"\n",
      "[1] \"model/ada/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.967\"\n",
      "[1] \"model/ada/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.993\"\n",
      "[1] \"model/ada/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.971\"\n",
      "[1] \"lda\"\n",
      "[1] \"model/lda/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.805\"\n",
      "[1] \"model/lda/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.961\"\n",
      "[1] \"model/lda/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.978\"\n",
      "[1] \"model/lda/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.957\"\n",
      "[1] \"knn\"\n",
      "[1] \"model/knn/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.832\"\n",
      "[1] \"model/knn/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.954\"\n",
      "[1] \"model/knn/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.969\"\n",
      "[1] \"model/knn/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.955\"\n",
      "[1] \"glm\"\n",
      "[1] \"model/glm/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.805\"\n",
      "[1] \"model/glm/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in predict.lm(object, newdata, se.fit, scale = 1, type = if (type == :\n",
      "\"prediction from rank-deficient fit; attr(*, \"non-estim\") has doubtful cases\"\n",
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.964\"\n",
      "[1] \"model/glm/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.975\"\n",
      "[1] \"model/glm/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.891\"\n",
      "[1] \"svmRadial\"\n",
      "[1] \"model/svmRadial/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.795\"\n",
      "[1] \"model/svmRadial/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.942\"\n",
      "[1] \"model/svmRadial/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.966\"\n",
      "[1] \"model/svmRadial/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.950\"\n",
      "[1] \"naive_bayes\"\n",
      "[1] \"model/naive_bayes/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.804\"\n",
      "[1] \"model/naive_bayes/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.962\"\n",
      "[1] \"model/naive_bayes/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.982\"\n",
      "[1] \"model/naive_bayes/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.978\"\n",
      "[1] \"xgbTree\"\n",
      "[1] \"model/xgbTree/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.836\"\n",
      "[1] \"model/xgbTree/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.966\"\n",
      "[1] \"model/xgbTree/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.992\"\n",
      "[1] \"model/xgbTree/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.972\"\n",
      "[1] \"treebag\"\n",
      "[1] \"model/treebag/clustering_log2FC_0.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.833\"\n",
      "[1] \"model/treebag/clustering_log2FC_1.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.965\"\n",
      "[1] \"model/treebag/clustering_log2FC_1.5.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.983\"\n",
      "[1] \"model/treebag/clustering_log2FC_2.RData\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = NRG, case = RG\n",
      "\n",
      "Setting direction: controls < cases\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AUC-ROC: 0.966\"\n"
     ]
    }
   ],
   "source": [
    "Results <- NULL\n",
    "evaluation <- NULL\n",
    "for(ML in c('rf','rpart','ada','lda','knn','glm','svmRadial','naive_bayes','xgbTree','treebag')){\n",
    "    print(ML)\n",
    "    roclist <- NULL\n",
    "    for(type in c('clustering')){\n",
    "        AUC <- c()\n",
    "        for(lfc in c(0.5,1,1.5,2)){\n",
    "            f <- paste0(\"model/\",ML,\"/\",type,\"_log2FC_\",lfc,\".RData\")\n",
    "            print(f)\n",
    "            load(f)\n",
    "            data <- rbind(train,test)\n",
    "            if(ML %in% c(\"svmRadial\")){\n",
    "                rf_pred <- predict(model,test)\n",
    "                rf_pred <- as.numeric(rf_pred)-1\n",
    "            }else{\n",
    "                rf_pred <- predict(model,test,type=\"prob\")\n",
    "                rf_pred <- rf_pred[,2]\n",
    "            }\n",
    "            RE <- metrics2(rf_pred,test$class)\n",
    "            RE$model <- ML\n",
    "            RE$log2FC <- lfc\n",
    "            evaluation <- rbind(evaluation,RE)\n",
    "            predob <- prediction(rf_pred, test$class)\n",
    "            perf <- performance(predob, 'tpr','fpr')\n",
    "            perf.auc<- performance(predob, measure = 'auc', x.measure = 'cutoff')\n",
    "            re_auc <- data.frame(Model=ML,FC=lfc,`AUC-ROC`=unlist(perf.auc@y.values))\n",
    "            Auc <- paste0(\"AUC-ROC: \",sprintf(\"%.3f\",unlist(perf.auc@y.values)))\n",
    "            print(Auc)\n",
    "            Results <- rbind(Results,re_auc)\n",
    "            AUC <- c(AUC,Auc)\n",
    "            re <- data.frame(x = attributes(perf)$x.values[[1]],\n",
    "                             y = attributes(perf)$y.values[[1]],\n",
    "                            type = str_to_title(type),\n",
    "                            lfc)\n",
    "            roclist <- rbind(roclist,re)\n",
    "        }\n",
    "        AUC <- paste(AUC,collapse = '\\n')\n",
    "    }    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9be3180",
   "metadata": {},
   "source": [
    "## Figure2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4928873e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtAAAALQCAMAAACOibeuAAAASFBMVEUAAAAzMzNBx9VFy9lI\njr9MksNSr1JWs1aVlZWZmZmfdGujeG+lgcaphcrFxkrJyk7aTk/eUlPljsrpks7r6+v7lTr/\nmT7///8z/gSZAAAACXBIWXMAABJ0AAASdAHeZh94AAAVtElEQVR4nO3djXbbuHaAUSZN0zRN\nc9PUqd//TTszluQD0LShA0okqP2tdWc8ujkCTO/Q1I+T6Vk6UNPWG5DWDGgdKqB1qIDWoQJa\nhwpoHSqgdaiA1qHqA/1J2kVrgW76VU9da2w+b/tbzreNAz3O8rbfENDjLG/7DQE9zvK23xDQ\n4yxv+w0BPc7ytt8Q0OMsb/sNAT3O8rbfENDjLG/7DQE9zvK231A76Mv/P/3V679m99O/p93O\n2/6W8yuDvuidXv4xlTNA73/+IbbfCnp6Bnrr5W2/oesvOeag/3mT05O0g9YAXf7GeK+HOEfs\ndv4htg/0OMvbfkNAj7O87TcE9DjL235DQI+zvO03dCXo8IqKF1ZGm3+I7Xvpe5zlbb8hoMdZ\n3vYbAnqc5W2/IaDHWd72GwJ6nOVtvyGgx1ne9hsCepzlt9v+FOocv/U80CMs3yuic/nzLrad\nbxsHepzle0UAfUVA32E+/SVa5wwP9KytRTwq6FXGt54HevX5rbcP9McB3dQhHpUBfUUHB/1S\n71cU6NuPA31FQG85D/Tq8x1PEKxxxQJ0Q0Bf0UOI2O080KvPP4SI3c4Dvfr8Q4jY7TzQoUO8\nVAZ0Qw8C+hQRA88DPY+IgeePBdrzXv3zD7H9UUC/NMQh3e38SNv/t3nT/KY3BoEeZ/lH2j7Q\n95gfScTGy//rjab5TYvzQN9jHujmgG7qgUTcYP6q8f+cN81vWhwHuqmRRGw8/7/zpvlNy/NA\ntwT03eaBBnrvywMN9Lrzd13+v+ZN85sWx4EGel/LAw30reevGv/vedP8puV5oIFedf5/5k3z\nm5bngQb6w4AGGujkPNBAfxjQQAMdAhpooJPzQAP9YUADDXToqj3+x7xpftPiONC7AV38zD7Q\nQI8O+qXPztBAA53tBqD/b940v2l5Hmig8wFdB/Sr5BjQQI8O+tRPZ2ig1wFdnCKBbgnouj2B\nful3+gwdfz8AHQJ6TNAvfXOGrgMaaKCPADp+zwYa6OFBv/TDGRpooIEG+p8OBLp42g1ooEcH\nfeq7MzTQQAOd6eCgi+/5QAN9aljQL/1yhga6CGiggQYaaKCfgQa6p09Pq/dn3jS/aXn+67xp\nftPi+K+mlpf/0dTy/PemFsd/N7W8/LemFsd/trU4/7mpNwYf5AxdPEvhDO0M/WE7B33qq0sO\noNu6Neh4hgUa6M1BF9/z2+7yjTP0H2dooPcBOrEG0EADHQIa6HxA1wENdBnQQAMdAhrofEDX\nAQ10GdBAAx0CGuh8QNcBDXQZ0EADHQIa6HxA1wENdBnQQAMdAhrofEDXAQ10GdBAAx0CGuh8\nQNcBDXQZ0EADHQIa6HxvgP73edMbty3eJdBAAx0CGuh8QNcBDXQZ0EADHQIa6HxA1wENdBnQ\nQAMdAhrofEDXAQ10GdBAAx0CGuh8QNcBDXQZ0EADHQIa6HxA1wENdBnQQAMdAhrofEDXAQ10\nGdBAAx0CGuh8QNcBfWjQxd8kCzTQo4N+6YszNNBAAw000EAD3RPQdUADXQY00ECHgAY6H9B1\nQANdBjTQQIeABjof0HVAA10GNNBAh4AGOh/QdUADXQY00ECHgAY6H9B1QANdBjTQQIeABnq5\nv39y6vWjvz+eXm96Bnoe0HsGPZW/Ypr9eqDrgAa6DGig7wR67hnoWUAPBvr1EvrT3z3N+tLW\nfPDUn6YWx5++NrU4/qup5eV/NLU8/72pxfHfTS0v/62pxfGfbS3Of27qjcEM6Gl2izP0PGfo\nYc7Q5bV0fT+XgAZ6CNBTdXN1P5eABnog0C45gB4Y9PmFlfD8RvG6CtCzgN416I8Cug5ooMuA\nBhroENBA5wO6Dmigy4AGGugQ0EDnA7oOaKDLgAYa6BDQQOcDug5ooMuABhroENBA5wO6Dmig\ny4AGGugQ0EDnA7oOaKDLgAYa6BDQQOcDug5ooMuABhroENBA5wO6Dmigy4AGGugQ0EDnA7oO\naKDLgAYa6BDQQOcDug5ooMuABhroENBA5wO6Dmigy4AGGugQ0EDnA7oOaKDLgAYa6BDQQOcD\nug5ooMuABhroENBA5wO6Dmigy4AGGugQ0EDnA7oOaKDLgAYa6BDQQOcDug5ooMuABhroENBA\n5wO6Dmigy4AGGugQ0EDnA7oOaKDLgAYa6BDQQOcDug5ooMuABhroENBA5wO6Dmigy4AGGugQ\n0EDnA7oOaKDLgAYa6BDQQOf79DTrS1vzwVN/mlocf/ra1OL4r6aWl//R1PL896YWx383tbz8\nt6YWx3+2tTj/uak3Bp2hl3KGfuwz9PwmoIEGOgQ00ECHgAY6H9B1QANdBjTQQIeABjof0HVA\nA10GNNBAh4AGOh/QdUADXQY00ECHgAY6H9B1QANdBjTQQIeABjof0HVAA10GNNBAh4AGOh/Q\ndUADXQY00ECHgAY6H9B1QANdBjTQQIeABjof0HVAA10GNNBAh4AGOh/QdUADXQY00ECHgAY6\nH9B1QANdBjTQQIeABjof0HVAA10GNNBAh4AGOh/QdUADXQY00ECHgAY6H9B1QANdBjTQQIeA\nBjof0HVAA10GNNBAh4AGOh/QdUADXQY00ECHgAY6H9B1QANdBjTQQIeABjof0HVAA10GNNBA\nh4AGOh/QdUADXQY00ECHgAY6H9B1QANdBjTQQIeABvq5/K8rjANdB/SeQE9v3PZuQNcBvSPQ\n01s3vhvQdUDvGvT0V+HDqbjlGeh5QO8ZdLyknma3PAM9D+gdgZ5dQwMN9NCgX64rpvL/nOKv\nAxrokUBXRdCnS+jXWz793dOsL23NB0/9aWpx/OlrU4vjv5paXv5HU8vz35taHP/d1PLy35pa\nHP/Z1uL856beGEyAfvmHM7Qz9Ehn6OVLjtN/Aw30QKDfe1D48iHQQI8D+r2n7VxyAD0+6PPL\nKFP40AsrQC+O7x70RwFdBzTQZUADvRZo77YD+lig66ftPgzoOqB3BfragK4DGugyoIFeDfT5\nesODQqCPAPr84okHhUAfCPQ1FyFA1wG9O9DLn+1793MJaKCBDgENNNAhoIG+fHxp+TNeup9L\nQAO9F9DXB3Qd0ECXAQ000CGggc4HdB3QQJcBDTTQIaCBzgd0HdBAlwENNNAhoIHOB3Qd0ECX\nAQ000CGggc4HdB3QQJcBDTTQIaCBzgd0HdBAlwENNNAhoIHOB3Qd0ECXAQ000CGggc4HdB3Q\nQJcBDTTQIaCBzgd0HdBAlwENNNAhoIHOB3Qd0ECXAQ000CGggc4HdB3QQJcBDTTQIaCBzgd0\nHdBAlwENNNAhoIHOB3Qd0ECXAQ000CGggc4HdB3QQJcBDTTQIaCBzgd0HdBAlwENNNAhoIHO\nB3Qd0ECXAQ000CGggc4HdB3QQJcBDfSGoJ9mfWlrPnjqT1OL409fm1oc/9XU8vI/mlqe/97U\n4vjvppaX/9bU4vjPthbnPzf1xqAz9FLO0I99hp7fBDTQQIeABhroENBA5wO6Dmigy4AGGugQ\n0EDnA7oOaKDLgAYa6BDQQOcDug5ooMuABhroENBA5wO6Dmigy4AGGugQ0EDnA7oOaKDLgAYa\n6BDQQOcDug5ooMuABhroENBA5wO6Dmigy4AGGugQ0EDnA7oOaKDLgAYa6BDQQOcDug5ooMuA\nBhroENBA5wO6Dmigy4AGGugQ0EDnA7oOaKDLgAYa6BDQQOcDug5ooMuABhroENBA5wO6Dmig\ny4AGGugQ0EDnA7oOaKDLgAYa6BDQQOcDug5ooMuABhroENBA5wO6Dmigy4AGGugQ0EDnA7oO\naKDLgAYa6BDQQOcDug5ooMuABhroENBA5wO6Dmigy4AGGugQ0EDnA7oOaKDLgAYa6BDQQC83\n/VX54RRuAnoe0HsGPb3+ivOH5a8Hug7oQUCf/xtooI8EuvrlQNcBPRjo10voT3/3NOtLW/PB\nU3+aWhx/+trU4vivppaX/9HU8vz3phbHfze1vPy3phbHf7a1OP+5qTcGc6Cn2S3O0HXO0OOc\noecfAD0L6GFAT+XN1f1cAhroIUBPs1uegZ4H9J5BX15NOb2gMhUvtTwDPQ/oXYP+KKDrgAa6\nDGiggQ4BDXQ+oOuABroMaKCBDgENdD6g64AGugxooIEOAQ10PqDrgAa6DGiggQ4BDXQ+oOuA\nBroMaKCBDgENdD6g64AGugxooIEOAQ10PqDrgAa6DGiggQ4BDXQ+oOuABroMaKCBDgENdD6g\n64AGugxooIEOAQ10PqDrgAa6DGiggQ4BDXQ+oOuABroMaKCBDgENdD6g64AGugxooIEOAQ10\nPqDrgAa6DGiggQ4BDXQ+oOuABroMaKCBDgENdD6g64AGugxooIEOAQ10PqDrgAa6DGiggQ4B\nDXQ+oOuABroMaKCBDgENdD6g64AGugxooIEOAQ10PqDrgAa6DGiggQ4BDXQ+oOuABroMaKCB\nDgENdD6g64AGugxooIEOAQ10vk9Ps760NR889aepxfGnr00tjv9qann5H00tz39vanH8d1PL\ny39ranH8Z1uL85+bemPQGXopZ+jHPkPPbwIaaKBDQAMNdAhooPMBXQc00GVAAw10CGig8wFd\nBzTQZUADDXQIaKDzAV0HNNBlQAMNdAhooPMBXQc00GVAAw10CGig8wFdBzTQZUADDXQIaKDz\nAV0HNNBlQAMNdAhooPMBXQc00GVAAw10CGig8wFdBzTQZUADDXQIaKDzAV0HNNBlQAMNdAho\noPMBXQc00GVAAw10CGig8wFdBzTQZUADDXQIaKDzAV0HNNBlQAMNdAhooPMBXQc00GVAAw10\nCGig8wFdBzTQZUADDXQIaKDzAV0HNNBlQAMNdAhooPMBXQc00GVAAw10CGig8wFdBzTQZUAD\nDXQIaKDzAV0HNNBlQAMNdAhooPMBXQc00GVAAw10CGig8wFdBzTQZUADDXQIaKCXm/6q/DDc\n8gz0PKD3DHp6/RWnD8Mt5f1cAhpooENAAw10CGigF3sX9Ke/e5J20A3P0G/01PSrdjtv+1vO\nt40DPc7ytt8Q0OMsb/sNAT3O8rbf0JUvrIRXVD58YSW9p93O2/6W8+uCbr+f93qIQ7rb+YfY\nPtDjLG/7DQE9zvK23xDQ4yxv+w0BPc7ytt8Q0OMsb/sNAT3O8rbfENDjLG/7DQE9zvK23xDQ\n4yxv+w0BPc7ytt8Q0OMsb/sNAT3O8rbfENDjLG/7DQE9zvK23xDQ4yxv+w0BPc7ytt8Q0OMs\nb/sNAT3O8rbfENDjLG/7Da0GWtpFK4FuZD/2vO1vOX/tONB7X972rwrovS9v+1cF9N6Xt/2r\nugdo6W4BrUMFtA4V0DpUQOtQAa1DBbT23ZVC7wN62982W/+m7Vx/6t9/zz30rn7n+TuB7vqi\n9E33jvcfou7td5vY8Oh3f/Vu/OuT9XxSUzeJvjPUtqDOfyXIRsv3j3f+drr5QKa/PqX8Z1X9\nPUW5u+g8ptt9TV8++zuf5arh7tNJfv706V87ceNOf9tQcvjyj44NdN3D/O9KumZ4uv5rUo9P\nfft/7vvsO38/d81P10/fAfRpQzkSp09pSop65dQlOv1VmV7LLv76F5F1zPd8h5h6roO7568d\nvcsZ+vSv9DkufZI7X+v0XYj2fTk7LhrC592xg67fzyeNfUcgOZ84P9/rGrpvuvv0kt1A5/VC\n9/frvu8v598IPY8gXr63bDM/pY7frUF3fsM83Ud28Py/3D1M/dtfSXT+cmvq/u7UNd4xf/nG\ntKtr6K7v95ehbtH5M1zikJZ3stY5OrVwH8jz5XfHI9r8fPiy7elZjp7veOG8tMVTJFP4d/oK\nMC+693LncsGRf4Lm+fQgpOMApufTTwbcHHTPY9yVXtHIH9F/Psgu3fX8RvfTI+Fvac/dQfcT\npp3zyScDbv6gsOMRdu933Oe+T6/nLl4vVDqewc6Ph+VTa1/u5L4vilTzfdep69d7CfZyH+lj\nknrSZ7Z41yOajR4Lv3yn7xa93ePBeAdXD3Us+PE991yCne4ld4Y6n906X/VNf3tJPJp5e/n0\nbN/8y7NtXRc8XfPne0nM9Cz47h33fscr7ufakfPlQue3ra4HVM/jij4PbzV/uZ/EF79zyeX7\nXeO7bnbtzvnOp1fCV3Ij0WdMWzw7tMZ8Rzda9XJEb3P376/9soHeS/eO4RW+23Y1dT4W3/zx\nYEe3WbTzAX7n4s/93+/6ngR+fRK1YwsdTR2PH6bLmb3jyZkVXh7Odqsz9C3v/MPFV7h+73uK\nZMOz8/P5t1L6uaHpueN3Y+98d7e7hr7lvV+xhfT0dtcs3fW9gaLnS9c739/tnuW47d03bKBz\n7fs/47RK5+9OHW+g6Hq77GZXmnELg91zw9rbHtSXLWyz7PkxRMrz+UKr77HDtuex7ZbWDZrS\nT76/jHe+X7Z3vj+gD1XvG/I7Xg+bOudXCuhjtcobKBLzl4u8ja84gD5Q52/3Pe/myj6mez0x\nb/zwBejDdDo3d72f/58Psp47X45aJ6CP0Pn8mH11sG+8uIet234HWqHTtUb6+bq+8dfT+9ZP\nlgJ9lDpPkT3j8WHo9py234FWaYXHg9nx8Hr39py234H6ujxf1n3B0Pfj4dMuPAM9euGHEfL3\nscJ478/ardUuNqF805sfJu4j+3bRnRna1WZ0fVPxr9x0z9tNu/7wkBu0n50oVdfzC1N4dbDn\nkmNHpneyDWXre8Ls8kiu9+0fu2lPe1GmKfwzNdx/dt0Toj3tRVd3frKt42dMVni+bU+I9rQX\nXdvrT6T23UPvNvrvYrX2tBddWe97n59Xuv7dE6I97UVX1v1wrn98dx3t83mo1nj78a6eolih\ng306j9X5LUV9d7LOXvbSwT6dx+nyx435GhY5GGM2Vf/TKQdjuF5/MmSNK46j5WgMV3grkfPz\nLIdjwDheznEZMaIXc1iGy2PB93JURmvzPw5x3zkwg+X8/H6Oy2Ct8f6NI+fQDNYa7984co7M\nSE3P67x/48A5MAPl/Rsf58CMk/dvNOTAjJP3bzTkyIzT6+NBX7XFHJpx8niwIUdnjDwebMzR\nGSKPB1tzdIZoOl9ruOL4IIdniF7/AhRfsPdzfHbe+VVulttykPbd5Y/5Sv/9KQ+Wg7Trwo8P\n8tyUo7TnTi+lsNyeQ7XjLn9Srq9Scw7Vrpv8wNWVOVT7rv8vjHiwHKqd54rjuhyrvecx4VU5\nVruP52tysHSogNahAlqHCmgdKqB1qIDWoQJahwpoHSqgdaiAvndT9e/6v3xFunL47t35R6qA\nvkkO370D+qY5fPcu/vkapzfSTfG/fEW6cvjuXQB98ht+zsqbn3tz+O7d9Ko5EAZ6pRy+e7cM\n2p/FuEIO37179wz97CvSmcN378IfIuqSY/0cvntX/Mm452c5Js9yrJTDp0MFtA4V0DpUQOtQ\nAa1DBbQOFdA6VEDrUAGtQwW0DhXQOlRA61D9PzQ97G06AvCnAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 360,
       "width": 360
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=6, repr.plot.height=6)\n",
    "Figure2C <- Results %>%\n",
    "    group_by(Model) %>%\n",
    "    summarise(ROC=mean(AUC.ROC),se=std.error(AUC.ROC)) %>%\n",
    "    arrange(ROC) %>%\n",
    "    mutate(Model=factor(Model,levels=c('rpart','glm','svmRadial','lda','knn','naive_bayes','treebag','ada','xgbTree','rf'))) %>%\n",
    "    mutate(Model=factor(Model,labels=c('rpart','LR','SVM','LDA','KNN','NB','BaggingClassifier','AdaBoost','XGBoost','RF'))) %>%\n",
    "    ggplot(aes(Model,ROC,fill=Model))+\n",
    "                geom_bar(stat='identity',width=0.6,alpha=0.8)+\n",
    "                geom_errorbar(aes(ymin = ROC - se, ymax = ROC + se),position = position_dodge(0.9), width = .2)+\n",
    "                theme_bw()+\n",
    "                theme(axis.text = element_text(size=8,color='black'),\n",
    "                      axis.text.x = element_text(size=8,color='black',angle=45,hjust=1),\n",
    "                      axis.title=element_text(size=8,color='black'),\n",
    "                     legend.position='none',\n",
    "                     strip.text = element_text(size = 8, colour = \"black\"))+\n",
    "scale_fill_d3()\n",
    "Figure2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15f57c08",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
